services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,528 @@
+    }
+
+    @Override
+    public Future<TenantResult<JsonObject>> get(final X500Principal subjectDn, final Span span) {
+        Objects.requireNonNull(subjectDn);
+        Objects.requireNonNull(span);
+
+        return getForCertificateAuthority(subjectDn, span);
+    }
+
+    /**
+     * Finds an existing tenant by the subject DN of its configured certificate authority.
+     *
+     * @param subjectDn the subject DN to find the tenant with.
+     * @param span The active OpenTracing span for this operation. It is not to be closed in this method!
+     *            An implementation should log (error) events on this span and it may set tags
+     *            and use this span as the parent for any spans created in this method.
+     * @return A future indicating the outcome of the operation.
+     *         The <em>status</em> will be
+     *         <ul>
+     *         <li><em>200 OK</em> if a tenant with the given certificate authority is registered.
+     *         The <em>payload</em> will contain the tenant's configuration information.</li>
+     *         <li><em>400 Bad Request</em> if the certificate authority is missing.</li>
+     *         <li><em>404 Not Found</em> if no tenant with the given identifier.</li>
+     *         </ul>
+     * @throws NullPointerException if the parameters are {@code null}.
+     */
+    private Future<TenantResult<JsonObject>> getForCertificateAuthority(final X500Principal subjectDn,
+            final Span span) {
+        Objects.requireNonNull(span);
+
+        if (subjectDn == null) {
+            TracingHelper.logError(span, "missing subject DN");
+            return Future.succeededFuture(TenantResult.from(HttpURLConnection.HTTP_BAD_REQUEST));
+        } else {
+            return findTenant(subjectDn)
+                    .compose(tenantDtoResult -> Future.succeededFuture(TenantResult.from(
+                            HttpURLConnection.HTTP_OK,
+                            DeviceRegistryUtils.convertTenant(tenantDtoResult.getTenantId(),
+                                    tenantDtoResult.getTenant(), true),
+                            DeviceRegistryUtils.getCacheDirective(config.getCacheMaxAge()))))
+                    .recover(error -> {
+                        TracingHelper.logError(span, "no tenant found for subject DN");
+                        return Future.succeededFuture(TenantResult.from(HttpURLConnection.HTTP_NOT_FOUND));
+                    });
+        }
+    }
+
+    @Override
+    public Future<OperationResult<Tenant>> readTenant(final String tenantId, final Span span) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(span);
+
+        return processReadTenant(tenantId, span)
+                .recover(error -> Future.succeededFuture(MongoDbDeviceRegistryUtils.mapErrorToResult(error, span)));
+    }
+
+    private Future<OperationResult<Tenant>> processReadTenant(final String tenantId, final Span span) {
```

Comment:
The `span` here remains unused, in that case it could be removed.

Reply:
done